File: javav2/example_code/cognito/src/main/java/com/example/cognito/AddLoginProvider.java
//snippet-sourcedescription:[AddLoginProvider.java demonstrates how to associate an Amazon Cognito identity pool with an identity provider.]
//snippet-keyword:[AWS SDK for Java v2]
//snippet-keyword:[Amazon Cognito]

/*
   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
   SPDX-License-Identifier: Apache-2.0
*/

package com.example.cognito;

//snippet-start:[cognito.java2.add_login_provider.import]
import software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.cognitoidentity.CognitoIdentityClient;
import software.amazon.awssdk.services.cognitoidentity.model.CognitoIdentityProvider;
import software.amazon.awssdk.services.cognitoidentity.model.UpdateIdentityPoolRequest;
import software.amazon.awssdk.services.cognitoidentity.model.UpdateIdentityPoolResponse;
import software.amazon.awssdk.services.cognitoidentityprovider.model.CognitoIdentityProviderException;
import java.util.ArrayList;
import java.util.List;
//snippet-end:[cognito.java2.add_login_provider.import]

/**
 * Before running this Java V2 code example, set up your development environment, including your credentials.
 *
 * For more information, see the following documentation topic:
 *
 * https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/get-started.html
 */
public class AddLoginProvider {

    public static void main(String[] args) {

        final String usage = "\n" +
            "Usage:\n" +
            "    <appId> <identityPoolName> <identityPoolId> <providerName>\n\n" +
            "Where:\n" +
            "    appId - The application ID value from the login provider.\n\n" +
            "    identityPoolName - The name of your identity pool.\n\n" +
            "    identityPoolId - The Id value of your identity pool.\n\n" +
            "    providerName - The provider name (i.e., cognito-idp.us-east-1.amazonaws.com/us-east-1_Taz4Yxxxx).\n\n";

        if (args.length != 4) {
            System.out.println(usage);
            System.exit(1);
        }

        String appId = args[0];
        String identityPoolName = args[1];
        String identityPoolId = args[2];
        String providerName = args[3];

        CognitoIdentityClient cognitoclient = CognitoIdentityClient.builder()
            .region(Region.US_EAST_1)
            .credentialsProvider(ProfileCredentialsProvider.create())
            .build();

        setLoginProvider(cognitoclient,appId, identityPoolName, identityPoolId, providerName) ;
        cognitoclient.close();
    }

    //snippet-start:[cognito.java2.add_login_provider.main]
    public static void setLoginProvider(CognitoIdentityClient cognitoclient,
                                        String appId,
                                        String identityPoolName,
                                        String identityPoolId,
                                        String providerName) {

        CognitoIdentityProvider identityProvider = CognitoIdentityProvider.builder()
            .providerName(providerName)
            .clientId(appId)
            .build();

        List<CognitoIdentityProvider> proList = new ArrayList<>();
        proList.add(identityProvider);
        try {
            UpdateIdentityPoolRequest poolRequest = UpdateIdentityPoolRequest.builder()
                .allowUnauthenticatedIdentities(true)
                .identityPoolName(identityPoolName)
                .identityPoolId(identityPoolId)
                .cognitoIdentityProviders(proList)
                .build() ;

            UpdateIdentityPoolResponse response = cognitoclient.updateIdentityPool(poolRequest);
            List<CognitoIdentityProvider> providers = response.cognitoIdentityProviders();
            for (CognitoIdentityProvider provider: providers) {
                System.out.println("The client ID is : "+provider.clientId());
                System.out.println("The provider name is : "+provider.providerName());
            }

        } catch (CognitoIdentityProviderException e){
            System.err.println(e.awsErrorDetails().errorMessage());
            System.exit(1);
        }
    }
    //snippet-end:[cognito.java2.add_login_provider.main]
}

File: php/example_code/cognito/AddLoginProvider.php
<?php
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 *  ABOUT THIS PHP SAMPLE: This sample is part of the SDK for PHP Developer Guide topic at
 * https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/iam-examples-working-with-policies.html
 *
 */
// snippet-start:[cognito.php.identity_pool.add_login_provider.complete]
// snippet-start:[cognito.php.identity_pool.add_login_provider.import]

require 'vendor/autoload.php';

use Aws\CognitoIdentity\CognitoIdentityClient;
use Aws\Exception\AwsException;

// snippet-end:[cognito.php.identity_pool.add_login_provider.import]

/**
 *  Adds a login provider to an Identity Pool for your AWS account.
 *
 * This code expects that you have AWS credentials set up per:
 * https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/guide_credentials.html
 */

 // snippet-start:[cognito.php.identity_pool.add_login_provider.main]

$identityClient = new CognitoIdentityClient([
    'profile' => 'default',
    'region' => 'us-east-2',
    'version' => '2014-06-30'
]);

$identityPoolName = "PHP_SDK_test_identity_pool";
$identityPoolid = "us-east-2:1234567890112-abcdefg";
$appid = "amzn1.application-oa2-client.1234567890112-abcdefg";

$facebook = "graph.facebook.com";
$google = "accounts.google.com";
$amazon = "www.amazon.com";
$twitter = "api.twitter.com";
$digits = "www.digits.com";


try {
    $result = $identityClient->updateIdentityPool([
        'AllowUnauthenticatedIdentities' => false,
        'IdentityPoolName' => $identityPoolName,
        'IdentityPoolId' => $identityPoolid,
        'SupportedLoginProviders' => [$amazon => $appid ]
    ]);
    echo $result["IdentityPoolId"]["ClientId"];
    var_dump($result);
} catch (AwsException $e) {
    // output error message if fails
    echo $e->getMessage() . "\n";
    error_log($e->getMessage());
}

// snippet-end:[cognito.php.identity_pool.add_login_provider.main]
// snippet-end:[cognito.php.identity_pool.add_login_provider.complete]
// snippet-sourceauthor:[jschwarzwalder (AWS)]
